pkg/server/cloud/gce/instances.go

```diff
@@ -0,0 +1,445 @@
+		size = minimumDiskSize
+	}
+	diskSpec := []*compute.AttachedDisk{
+		{
+			Boot:       isBoot,
+			AutoDelete: true,
+			// RW or RO
+			Mode: "READ_WRITE",
+			// SCSI or NVMe
+			Interface: "SCSI",
+			InitializeParams: &compute.AttachedDiskInitializeParams{
+				DiskName:    name,
+				DiskSizeGb:  size,
+				DiskType:    typeURL,
+				SourceImage: imageURL,
+			},
+		},
+	}
+	return diskSpec
+}
+
+func (c *gceClient) getInstanceNetworkSpec(privateIPOnly bool) []*compute.NetworkInterface {
+	networkURL := c.getNetworkURL()
+	subNetworkURL := c.getSubNetworkURL()
+
+	accessConfig := &compute.AccessConfig{
+		Name:        "External NAT",
+		NetworkTier: "STANDARD",
+		Type:        "ONE_TO_ONE_NAT",
+	}
+	// If we are only using private IP set the access config to nil
+	if privateIPOnly || !c.usePublicIPs {
+		accessConfig = nil
+	}
+	networkSpec := []*compute.NetworkInterface{
+		{
+			AccessConfigs: []*compute.AccessConfig{accessConfig},
+			AliasIpRanges: []*compute.AliasIpRange{
+				{
+					IpCidrRange: "/32",
+				},
+			},
+			Network:    networkURL,
+			Subnetwork: subNetworkURL,
+		},
+	}
+	return networkSpec
+}
+
+func (c *gceClient) createInstanceSpec(node *api.Node, metadata string) (*compute.Instance, error) {
+	md, err := base64.StdEncoding.DecodeString(metadata)
+	if err != nil {
+		return nil, util.WrapError(err, "Could not decode metadata string")
+	}
+	mds := string(md)
+	name := makeInstanceID(c.controllerID, node.Name)
+	diskType := c.getDiskTypeURL()
+	volSizeGiB := cloud.ToSaneVolumeSize(node.Spec.Resources.VolumeSize)
+	disks := c.getAttachedDiskSpec(true, int64(volSizeGiB), name, diskType, node.Spec.BootImage)
+	labels := c.getInstanceLabels(node.Name)
+	networkInterfaces := c.getInstanceNetworkSpec(node.Spec.Resources.PrivateIPOnly)
+	instanceType := c.getInstanceTypeURL(node.Spec.InstanceType)
+	spec := &compute.Instance{
+		Disks:             disks,
+		Labels:            labels,
+		MachineType:       instanceType,
+		Name:              name,
+		NetworkInterfaces: networkInterfaces,
+		Tags: &compute.Tags{
+			Items: c.bootSecurityGroupIDs,
+		},
+		Metadata: &compute.Metadata{
+			Items: []*compute.MetadataItems{
+				{
+					Key:   "user-data",
+					Value: &mds,
+				},
+			},
+		},
+	}
+	if node.Spec.Spot {
+		ar := false
+		spec.Scheduling = &compute.Scheduling{
+			AutomaticRestart:  &ar,
+			OnHostMaintenance: "TERMINATE",
+			Preemptible:       true,
+		}
+	}
+	return spec, nil
+}
+
+func (c *gceClient) StartNode(node *api.Node, metadata string) (*cloud.StartNodeResult, error) {
+	klog.V(2).Infof("Starting instance for node: %v", node)
+	spec, err := c.createInstanceSpec(node, metadata)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("Starting node with security groups: %v subnet: '%s'",
+		c.bootSecurityGroupIDs, c.subnetName)
+	op, err := c.service.Instances.Insert(c.projectID, c.zone, spec).Do()
```

Comment:
It seems a context is used for timeouts everywhere else, any reason not to do the same here?

Reply:
fixed in d7bc5a7